### ISSUE TITLE
Using cy.server() and cy.route() not working in before hooks

### ISSUE THREAD

**Issue by pawelgalazka** [cypress-io/cypress#595]
- Cypress Version: 0.19.2

### Is this a Feature or Bug?

Bug

### Current behavior:

I'm getting error in Cypress app console: `CypressError: cy.route() cannot be invoked before starting the cy.server()` when I use `cy.server()` in top `before` hook and then `cy.route()` in children contexts `before` hooks.

If I remove top `cy.server()` and move it to `before` hook before each `cy.route()` call then I get fail for second describe context with error: `CypressError: The XHR server is unavailable or missing. This should never happen and likely is a bug. Open an issue if you see this message.` 

### Desired behavior:

I should be able to use `cy.route()` and `cy.server()` within `before` hooks.

#### Test code:

```javascript
describe('test', () => {
  before(() => {
    cy.server()
  })

  describe('case with http 500', () => {
    before(() => {
      // cy.server()
      cy.route({
        method: 'GET',
        url: '/api?*',
        status: 500,
        response: {}
      })
      cy.visit('http://app.dev/')
    })

    it('test n1', () => {

    })

    it('test nr 2', () => {

    })
  })

  describe('case with http 408', () => {
    before(() => {
      // cy.server()
      cy.route({
        method: 'GET',
        url: '/api?*',
        status: 408,
        response: {}
      })
      cy.visit('http://app.dev')
    })

    it('test nr 3', () => {

    })

    it('test nr 4', () => {

    })
  })
})
```

**Comment by jennifer-shehane** [comment]
Can confirm this is a bug in `0.19.4`.
<img width="626" alt="screen shot 2017-08-17 at 10 47 38 am" src="https://user-images.githubusercontent.com/1271364/29418152-8f3cf5e8-8339-11e7-9550-7380b3e63e32.png">

**Comment by jennifer-shehane** [comment]
The second use case has been fixed in the upcoming `0.20.0` version. 

The first use case still fails in the upcoming `0.20.0` version. 

**Minimum code to reproduce**
```js
describe('test', () => {
  before(() => {
    cy.server()
  })

  describe('describe n1', () => {
    before(() => {
      cy.route('fake')
    })

    it('test n1', () => {})
  })

  describe('describe n2', () => {
    before(() => {
      cy.route('fake')
    })

    it('test nr 2', () => {})
  })
})
```

**Comment by brian-mann** [comment]
I looked at this, and this is not a bug, this is the correct and intended behavior of mocha hooks.

Code in a `before` hook only runs **once** and once only. When it reaches the 2nd test: `test nr 2`  the `server` has been reset and needs to be reinvoked.

Replace the notion of `cy.server` with anything else in the before hook - whether its another cypress command, or other javascript code. It only makes sense to run once and will not affect nor have anything to do with the execution of `test nr 2`.

You just need to move the `before` hook to `beforeEach` since you intend it to run **before each** test and not just before the **first test**.

**Comment by brian-mann** [comment]
To clarify I only ran @jennifer-shehane's minimal code example in our upcoming `0.20.0`. I suppose something about `0.19.4` could be different, but I'm closing this issue for now. Can reopen if another another demonstrable example is buggy.

**Comment by brian-mann** [comment]
Fixed in [`0.20.0`](https://docs.cypress.io/guides/references/changelog.html#0-20-0)
